staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go

```diff
@@ -179,20 +179,41 @@ func (a *Authenticator) UpdateTransportConfig(c *transport.Config) error {
 		return &roundTripper{a, rt}
 	}
 
-	getCert := c.TLS.GetCert
-	c.TLS.GetCert = func() (*tls.Certificate, error) {
-		// If previous GetCert is present and returns a valid non-nil
-		// certificate, use that. Otherwise use cert from exec plugin.
-		if getCert != nil {
-			cert, err := getCert()
-			if err != nil {
-				return nil, err
+	if t, ok := c.Transport.(*http.Transport); ok && t != nil {
+		if t.TLSClientConfig == nil {
+			t.TLSClientConfig = &tls.Config{}
+		}
+		getCert := t.TLSClientConfig.GetClientCertificate
+		t.TLSClientConfig.GetClientCertificate = func(hi *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+			// If previous GetCert is present and returns a valid non-nil
```

Comment:
trying to think through the implications if you use client cert rotation *and* an exec plugin... the client cert rotation code would always win. does that make sense?

Reply:
where are we checking if the returned cert is valid?

Reply:
> where are we checking if the returned cert is valid?

`refreshCredsLocked` does validation. It fetches the actual certificate, validates and caches it. `GetClientCertificate` returns the cached copy.
I added cert expiry validation too in this PR, just in case.

> trying to think through the implications if you use client cert rotation and an exec plugin... the client cert rotation code would always win. does that make sense?

Is cert rotation *and* exec plugin a valid configuration? Does it mean "fetch initial cert via plugin but then rotate manually"?
Our use case relies on exec plugin to bootstrap a fresh cert whenever needed and cert rotation in kubelet is disabled.

Reply:
> refreshCredsLocked does validation.

I was referring to what the previous GetCert was returning ("If previous GetCert is present and returns a *valid* non-nil..."), not what the exec plugin returned

> Is cert rotation and exec plugin a valid configuration?

I don't know, but this is allowing it :)

Reply:
> I was referring to what the previous GetCert was returning ("If previous GetCert is present and returns a valid non-nil..."), not what the exec plugin returned

Ah, gotcha. I don't think it's this plugin's job to check that. If something returns a certificate here, it better be valid. Otherwise it's a bug.

I'd rather leave it as is, WDYT?

Reply:
For the sanity of the next person who tries to debug this, I think it would be better to return an error if GetClientCertificate is not nil. That would effectively disallow client cert rotation to be used with the exec plugin, right? I'd be happy to avoid this complexity.

Reply:
> I think it would be better to return an error if GetClientCertificate is not nil. That would effectively disallow client cert rotation to be used with the exec plugin, right? I'd be happy to avoid this complexity.

that seems like a reasonable starting point. mixing the two gets weird.

Reply:
SGTM, made it an error in both cases